awx_collection/plugins/lookup/tower_api.py

```diff
@@ -0,0 +1,202 @@
+# (c) 2020 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = """
+lookup: tower_api
+author: John Westcott IV (@john-westcott-iv)
+short_description: Search the API for objects
+requirements:
+  - None
+description:
+  - Returns GET requests from the Ansible Tower API. See
+    U(https://docs.ansible.com/ansible-tower/latest/html/towerapi/index.html) for API usage.
+  - For use that is cross-compatible between the awx.awx and ansible.tower collection
+    see the tower_meta module
+extends_documentation_fragment: awx.awx.auth_plugin
+options:
+  _terms:
+    description:
+      - The endpoint to query, i.e. teams, users, tokens, job_templates, etc.
+    required: True
+  query_params:
+    description:
+      - The query parameters to search for in the form of key/value pairs.
+    type: dict
+    required: False
+    aliases: [query, data, filter, params]
+  expect_objects:
+    description:
+      - Error if the response does not contain either a detail view or a list view.
+    type: boolean
+    default: False
+    aliases: [expect_object]
+  expect_one:
+    description:
+      - Error if the response contains more than one object.
+    type: boolean
+    default: False
+  return_objects:
+    description:
+      - If a list view is returned, promote the list of results to the top-level of list returned.
+      - Allows using this lookup plugin to loop over objects without additional work.
+    type: boolean
+    default: True
+  return_all:
+    description:
+      - If the response is paginated, return all pages.
+    type: boolean
+    default: False
+  return_ids:
+    description:
+      - If response contains objects, promote the id key to the top-level entries in the list.
+      - Allows looking up a related object and passing it as a parameter to another module.
+      - This will convert the return to a string or list of strings depending on the number of selected items.
+    type: boolean
+    aliases: [return_id]
+    default: False
+  max_objects:
+    description:
+      - if C(return_all) is true, this is the maximum of number of objects to return from the list.
+      - If a list view returns more an max_objects an exception will be raised
+    type: integer
+    default: 1000
+
+notes:
+  - If the query is not filtered properly this can cause a performance impact.
+  - In addition, the built in threshold is 10,000 items; if the query returns more an exception will be thrown.
+"""
+
+EXAMPLES = """
+- name: Load the UI settings
+  set_fact:
+    tower_settings: "{{ lookup('awx.awx.tower_api', 'settings/ui') }}"
+
+- name: Lookup any users who are admins
+  set_fact:
+    admin_user_list_view: "{{ lookup('awx.awx.tower_api', 'users', query_params={ 'is_superuser': true }, return_objects=False) }}"
```

Comment:
I really wanted to put together examples where the return type would be obvious from use. That's why I liked

```yaml
    - name: Lookup any users who are admins  # use query to return the full list
      debug:
        msg: "{{ query('awx.awx.tower_api', 'users', query_params={ 'is_superuser': true }) | map(attribute='username') }}"
```

But with what you have, the user has to trust that the naming `_list_view` is accurate (in addition to knowing what an AWX list view is to begin with).

And I also question if we need _an_ example of `return_objects=False` to begin with. It's just not useful in any way that I can image.

Reply:
I tweaked it slightly:
```
    - name: Report the usernames of all users with admin privs
      debug:
        msg: "Admin users: {{ query('awx.awx.tower_api', 'users', query_params={ 'is_superuser': true }) | map(attribute='username') | join(', ') }}"
```

Let me know if you are good with that.

Reply:
🥇 💯 🥳  is how I feel about that